pkg/tbtc/node.go

```diff
@@ -552,6 +643,165 @@ func (n *node) handleRedemptionProposal(
+		cls = settings[0]
+	default:
+		cls = &coordinationLayerSettings{
+			executeCoordinationProcedureFn: executeCoordinationProcedure,
+			processCoordinationResultFn:    processCoordinationResult,
+		}
+	}
+
+	blockCounter, err := n.chain.BlockCounter()
+	if err != nil {
+		return fmt.Errorf("cannot get block counter: [%w]", err)
+	}
+
+	coordinationResultChan := make(chan *coordinationResult)
+
+	// Prepare a callback function that will be called every time a new
+	// coordination window is detected.
+	onWindowFn := func(window *coordinationWindow) {
+		// Fetch all wallets controlled by the node. It is important to
+		// get the wallets every time the window is triggered as the
+		// node may have started controlling a new wallet in the meantime.
+		walletsPublicKeys := n.walletRegistry.getWalletsPublicKeys()
+
+		for _, currentWalletPublicKey := range walletsPublicKeys {
+			// Run an independent coordination procedure for the given wallet
+			// in a separate goroutine. The coordination result will be sent
+			// to the coordination result channel.
+			go func(walletPublicKey *ecdsa.PublicKey) {
+				result, ok := cls.executeCoordinationProcedureFn(
+					n,
+					window,
+					walletPublicKey,
+				)
+				if ok {
+					coordinationResultChan <- result
+				}
+			}(currentWalletPublicKey)
+		}
+	}
+
+	// Start the coordination windows watcher.
+	go watchCoordinationWindows(
+		ctx,
+		blockCounter.WatchBlocks,
+		onWindowFn,
+	)
+
+	// Start the coordination result processor.
+	go func() {
+		for {
+			select {
+			case result := <-coordinationResultChan:
+				go cls.processCoordinationResultFn(n, result)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// executeCoordinationProcedure executes the coordination procedure for the
+// given wallet and coordination window.
+func executeCoordinationProcedure(
+	node *node,
+	window *coordinationWindow,
+	walletPublicKey *ecdsa.PublicKey,
+) (*coordinationResult, bool) {
+	walletPublicKeyBytes, err := marshalPublicKey(walletPublicKey)
+	if err != nil {
+		logger.Errorf("cannot marshal wallet public key: [%v]", err)
+		return nil, false
+	}
+
+	procedureLogger := logger.With(
+		zap.Uint64("coordinationBlock", window.coordinationBlock),
+		zap.String("wallet", fmt.Sprintf("0x%x", walletPublicKeyBytes)),
+	)
+
+	procedureLogger.Infof("starting coordination procedure")
+
+	executor, ok, err := node.getCoordinationExecutor(walletPublicKey)
+	if err != nil {
+		procedureLogger.Errorf("cannot get coordination executor: [%v]", err)
+		return nil, false
+	}
+	// This check is actually redundant. We know the node controls some
+	// wallet signers as we just got the wallet from the registry.
+	// However, we are doing it just in case. The API contract of
+	// getWalletsPublicKeys and/or getCoordinationExecutor may change one day.
+	if !ok {
+		procedureLogger.Infof("node does not control signers of this wallet")
+		return nil, false
+	}
+
+	result, err := executor.coordinate(window)
```

Comment:
Where will the code check if the current client is the leader for this window?

Reply:
Inside `executor.coordinate`, see: https://github.com/keep-network/keep-core/pull/3744#discussion_r1405813388